### ISSUE TITLE
Assumptions behind assembly and resource names, namespaces and paths

### ISSUE THREAD

**Issue by PMJMendes** [xaviersolau/BlazorJsonLocalization#48]
(Using .NET 7)

I've been banging my head trying to get the embedded localization json files to load...

I have a project where the default namespace has no relation with the assembly name. As such, a type's FullName is decorated with a namespace that the ResourcePathHelper has no idea how to handle.

Also, you're replacing all the '.' with '/', but the EmbeddedFileProvider is a flat provider, and all the resources are given by accessing their complete type name in the root folder, so using '/'s doesn't seem to work at all...

So, in order for this to work at all with my setup, I had to:

1. change line 113 of EmbeddedJsonLocalizationExtensionService to: `return new EmbeddedFileProvider(assembly, "");`
2. comment out lines 39-45 of ResourcePathHelper
3. leave the resourcespath property empty and put the json files in the same folder as the razor components

Mind you, I quite like having done point 3, as it's just easier to refer to both the razor component and the json file when they're side by side, but not everyone is like me... ;) Also, there should be some way to fix it so that other people don't have to do 1 and 2.

If I follow the setup for having a dedicated json folder, the resources get stored in the assembly as somthething like:

`part.of.namespace.path.to.resources.rest.of.namespace.typename.json`

and there's no reliable generic way to get at them...

I propose you do the following changes:

1. Keep my change to EmbeddedJsonLocalizationExtensionService, with the "" argument to the EmbeddedFileProvider constructor
2. Add a DefaultNamespace option to EmbeddedJsonLocalizationOptions
3. In EmbeddedJsonLocalizationExtensionService.TryLoadAsync (lines 39-60), if DefaultNamespace is not null, take the ResourcesPath property, replace any '/' with '.', and insert the value of the property in between the default part of the type's full name and the specific part

Otherwise, you're basically forcing the client assembly's name to be the same as its project's default namespace, and you're forcing all its types to be somewhere in that namespace.

**Comment by xaviersolau** [comment]
Hi,
Thanks for this feedback!

What version of the project do you use?

I will provide a fix for this point.
I think I am going to add a new option in EmbeddedJsonLocalizationOptions:
```csharp
        /// <summary>
        /// Gets/Sets Assembly Root NameSpace resolver.
        /// </summary>
        /// <remarks>
        /// By default the root name space will be resolved as the assembly name.
        /// </remarks>
        public Func<Assembly, string?>? RootNameSpaceResolver { get; set; }
```

This way you will be able to resolve yourself the root name space for the given assembly.

**Comment by PMJMendes** [comment]
I downloaded your main branch. :)

The new property will certainly help, and the default is reasonable.

**Comment by PMJMendes** [comment]
One more thing I did: I named all my json files .razor.json and changed the extension service accordingly. That way, VS2022 lists them neatly under the associated component, as if it were a CSS file. :)

So, I might suggest you actually go larger and instead of (or in addition to) RootNameSpaceResolver, add a ResourceNameResolver:

```
public Func<Type, string?>? ResourceNameResolver{ get; set; }
```

**Comment by xaviersolau** [comment]
Ok I will add an option to handle this.

Didn't you get any trouble to load embedded json named with `*.razor.json` or `*.razor.en.json` or `*.razor.fr.json`?

The fact is that there are some limitations on the embedded file name with several "dot": see https://learn.microsoft.com/en-us/dotnet/core/resources/manifest-file-names

On my side I am experiencing issues with embedded file name like `*.razor.fr.json` and I have to name it like this `*.razor-fr.json`.

**Comment by PMJMendes** [comment]
> Didn't you get any trouble to load embedded json named with `*.razor.json` or `*.razor.en.json` or `*.razor.fr.json`?
> The fact is that there are some limitations on the embedded file name with several "dot": see https://learn.microsoft.com/en-us/dotnet/core/resources/manifest-file-names

They're working like a peach. :)

From what I gathered, these points are specifically about .resx files implicitly included by MSBuild. Files explicitly added added as embedded resources by the user (seem to) have no such limitations.

This article is also almost a year old, so this may well be a .NET6 vs .NET 7 issue, or it could even be an MSBuild vs VS2022 issue.

**Comment by xaviersolau** [comment]
Ok I think I have a solution to solve the dot issue. I will be back with a PR soon!...

**Comment by xaviersolau** [comment]
I updated the examples to show how to handle the new NamingPolicy option.

in project SoloX.BlazorJsonLocalization.Example.Components.Embedded3

branch dev

**Comment by xaviersolau** [comment]
Well it's merged in main.

**Comment by xaviersolau** [comment]
available in nuget version 2.0.0-alpha.4

**Comment by PMJMendes** [comment]
Works like a charm! :)

Here's the config I'm using:

```
builder.Services.AddJsonLocalization(builder => builder.UseEmbeddedJson(options =>
{
    options.RootNameSpaceResolver = _ => typeof(App).Namespace!;
    options.NamingPolicy = (path, culture) =>
        $"{path.Replace('/', '.')}.razor{(string.IsNullOrEmpty(culture) ? string.Empty : $"-{culture}")}.json";
}));
```

Guess you can close this issue. Thank you kindly for this! :)
